Backend/Remora.Discord.Gateway/BaseGatewayClient.cs

```diff
@@ -0,0 +1,572 @@
+    protected CancellationTokenSource DisconnectRequestedSource { get; private set; }
+
+    /// <inheritdoc />
+    public TimeSpan Latency { get; protected set; }
+
+    /// <inheritdoc />
+    public GatewayConnectionStatus Status { get; protected set; }
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="BaseGatewayClient"/> class.
+    /// </summary>
+    /// <param name="logger">The logging provider.</param>
+    /// <param name="transportService">The payload transport service.</param>
+    /// <param name="dispatchService">The gateway event dispatch service.</param>
+    protected BaseGatewayClient
+    (
+        ILogger<BaseGatewayClient> logger,
+        IPayloadTransportService transportService,
+        IResponderDispatchService dispatchService
+    )
+    {
+        _logger = logger;
+        _payloadSendQueue = Channel.CreateUnbounded<IPayload>();
+        _preShutdownCommands = new Dictionary<Guid, IGatewayCommand>();
+        this.DispatchService = dispatchService;
+        this.TransportService = transportService;
+
+        _dispatchCts = new CancellationTokenSource();
+        this.DisconnectRequestedSource = new CancellationTokenSource();
+    }
+
+    /// <inheritdoc />
+    public virtual async Task<Result> RunAsync(CancellationToken ct = default)
+    {
+        if (this.Status is not GatewayConnectionStatus.Offline)
+        {
+            return new InvalidOperationError("Already running.");
+        }
+
+        try
+        {
+            _dispatchCts.Dispose();
+            _dispatchCts = new CancellationTokenSource();
+
+            _responderDispatchTask?.Dispose();
+            _responderDispatchTask = this.DispatchService.RunAsync(_dispatchCts.Token);
+
+            while (!ct.IsCancellationRequested)
+            {
+                switch (this.Status)
+                {
+                    case GatewayConnectionStatus.Offline:
+                    case GatewayConnectionStatus.Disconnected:
+                    {
+                        // We might have switched state while running. This is a signal to stop and reconnect.
+                        if (_isRunning)
+                        {
+                            var stopResult = await StopAsync(true);
+                            if (!stopResult.IsSuccess)
+                            {
+                                return stopResult;
+                            }
+                        }
+
+                        _isRunning = true;
+                        this.DisconnectRequestedSource.Dispose();
+                        this.DisconnectRequestedSource = new CancellationTokenSource();
+
+                        var connectResult = await ConnectAsync(ct);
+                        if (!connectResult.IsSuccess)
+                        {
+                            _logger.LogWarning
+                            (
+                                "Failed to connect to the gateway. Connection will be reattempted\n{Error}",
```

Comment:
I think *may* is better grammar here, since a failure may lead to the gateway stopping entirely.